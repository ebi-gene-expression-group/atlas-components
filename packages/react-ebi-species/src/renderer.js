"use strict";

//*------------------------------------------------------------------*

var React = require('react');
var ReactDOM = require('react-dom');

//*------------------------------------------------------------------*

var Icon = require('./SpeciesIcon.jsx');

//*------------------------------------------------------------------*

module.exports = function(mountNode,species,colourOverride, colourPerKingdom) {
    ReactDOM.render(
        React.createElement(Icon,{
          species: species,
          colourOverride: colourOverride,
          colourPerKingdom: colourPerKingdom
        }), mountNode
    );
};
