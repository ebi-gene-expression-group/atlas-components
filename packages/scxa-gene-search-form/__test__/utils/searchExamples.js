const searchExamples = [
  {
    text: `Rick Sanchez`,
    url: `http://rickandmorty.wikia.com/wiki/Special:Search?query=Rick+Sanchez`
  },
  {
    text: `Unity`,
    url: `http://rickandmorty.wikia.com/wiki/Special:Search?query=Unity`
  },
  {
    text: `Galactic Federation`,
    url: `http://rickandmorty.wikia.com/wiki/Special:Search?query=Galactic+Federation`
  },
  {
    text: `Blemflarck`,
    url: `http://rickandmorty.wikia.com/wiki/Special:Search?query=Blemflarck`
  },
  {
    text: `anatomy`,
    url: `http://rickandmorty.wikia.com/wiki/Special:Search?query=anatomy`
  }
]

export default searchExamples
