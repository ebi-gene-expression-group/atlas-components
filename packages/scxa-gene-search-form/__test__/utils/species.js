const getRandomInt = (max) => Math.floor(Math.random() * Math.floor(max))

const species = [
  `Meeseek`,
  `Gromflomite`,
  `Cromulon`,
  `Zigerion`,
  `Moopian`,
  `Bliznarvian`,
  `Greebybobe`
]

const getAll = () => species
const getRandom = () => species[getRandomInt(species.length)]
const getRandomSet = () => species.filter(() => Math.random() > 0.5)

export { getAll, getRandom, getRandomSet }
