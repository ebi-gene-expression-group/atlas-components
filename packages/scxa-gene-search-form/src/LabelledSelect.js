import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

import ebiVfReactSelectReplacements from './ebiVfReactSelectReplacements'

const _onlyUnique = (value, index, self) => self.indexOf(value) === index

const _makeOption = (str) => {
  if (str === ``) {
    return {
      value: ``,
      label: `Any`
    }
  }

  return {
    value: str.trim(),
    label: str.trim()
  }
}

class LabelledSelect extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      value: props.value
    }

    this.onChange = this._onChange.bind(this)
  }

  _onChange(selectedOption) {
    this.setState({ value: selectedOption.value })
    this.props.onChange(selectedOption)
  }

  render() {
    const { name, topGroup, bottomGroup, bottomGroupLabel, onChange, statusMessage } = this.props

    const isDisabled = Boolean(statusMessage)

    const topOptions =
      topGroup
        .filter(_onlyUnique)
        .map(_makeOption)

    const bottomOptions = {
      label: bottomGroupLabel,
      options: [``].concat(bottomGroup)
        .filter(_onlyUnique)
        .sort()
        .map(_makeOption)
    }

    const selectedOption =
      topOptions.concat(bottomOptions.options).find(option => option.value === this.state.value) ||
      bottomOptions.options[0]

    const capitalisedName = name.charAt(0).toUpperCase() + name.slice(1)

    return (
      <div>
        <label htmlFor={name}>{capitalisedName}</label>
        <Select
          name={name}
          components={{ IndicatorSeparator: null, DropdownIndicator: ebiVfReactSelectReplacements.DropdownIndicator }}
          styles={ebiVfReactSelectReplacements.styles}
          inputId={`${name}-input`}
          isSearchable={false}
          onChange={this.onChange}
          options={topOptions.concat(bottomOptions)}
          isDisabled={isDisabled}
          placeholder={isDisabled ? statusMessage : null}
          value={isDisabled ? null : selectedOption} />
      </div>
    )
  }
}

LabelledSelect.propTypes = {
  name: PropTypes.string,
  topGroup: PropTypes.arrayOf(PropTypes.string),
  bottomGroup: PropTypes.arrayOf(PropTypes.string),
  bottomGroupLabel: PropTypes.string,
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  statusMessage: PropTypes.string,
}

LabelledSelect.defaultProps = {
  name: `Select`,
  topGroup: [],
  bottomGroup: [],
  bottomGroupLabel: `All`,
  statusMessage: null
}

export default LabelledSelect
