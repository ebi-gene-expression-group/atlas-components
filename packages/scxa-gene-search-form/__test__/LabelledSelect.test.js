import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import Select from 'react-select'
import LabelledSelect from '../src/LabelledSelect'

import * as species from './utils/species'

const props = {
  name: `Species`,
  topGroup: species.getRandomSet(),
  bottomGroup: species.getAll(),
  bottomGroupLabel: `All species`,
  statusMessage: null,
  onChange: ()=> {}
}

describe(`LabelledSelect`, () => {
  test(`is made up of a label and a React Select component`, () => {
    const wrapper = shallow(<LabelledSelect {...props} />)
    expect(wrapper.first().find(`label`)).toHaveLength(1)
    expect(wrapper.first().find(`label`).text()).toEqual(props.name)
    expect(wrapper.last().find(Select)).toHaveLength(1)
  })

  test(`is disabled when a statusMessage (i.e. an error message) is passed in and it is displayed`, () => {
    const errorMessage = `Any schwifty, non-empty string`
    const wrapper = shallow(<LabelledSelect {...props} statusMessage={errorMessage} />)
    expect(wrapper.find(Select).prop(`isDisabled`)).toEqual(true)
    expect(wrapper.find(Select).prop(`placeholder`)).toEqual(errorMessage)
  })

  test(`is enabled when statusMessage is null`, () => {
    const wrapper = shallow(<LabelledSelect {...props} />)
    expect(wrapper.find(Select).prop(`isDisabled`)).toEqual(false)
  })

  test(`selects Any by default`, () => {
    const wrapper = shallow(<LabelledSelect {...props} />)
    expect(wrapper.find(Select).prop(`value`)).toHaveProperty(`value`, ``)
    expect(wrapper.find(Select).prop(`value`)).toHaveProperty(`label`, `Any`)
  })

  test(`can select an option via props`, () => {
    const selectedSpecies = species.getRandom()
    const wrapper = shallow(<LabelledSelect {...props} value={selectedSpecies}/>)
    expect(wrapper.find(Select).prop(`value`)).toHaveProperty(`value`, selectedSpecies)
  })

  test(`selects Any if value doesnâ€™t match any in option in topGroup or bottomGroup`, () => {
    const wrapper = shallow(<LabelledSelect {...props} selectedSpecies={`Human`}/>)
    expect(wrapper.find(Select).prop(`value`)).toHaveProperty(`value`, ``)
    expect(wrapper.find(Select).prop(`value`)).toHaveProperty(`label`, `Any`)
  })

  test(`changing the value in the React Select component fires the onChange prop function`, () => {
    const onChangeMock = jest.fn()
    const wrapper = shallow(<LabelledSelect {...props} onChange={onChangeMock} />)

    const speciesToSelect = species.getRandom()
    wrapper.find(Select).simulate(`change`, { value: speciesToSelect, label: speciesToSelect })

    expect(onChangeMock.mock.calls).toHaveLength(1)
    expect(onChangeMock.mock.calls[0][0]).toEqual({ value: speciesToSelect, label: speciesToSelect })
  })

  test(`matches snapshot`, () => {
    const tree = renderer.create(<LabelledSelect {...props}/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
