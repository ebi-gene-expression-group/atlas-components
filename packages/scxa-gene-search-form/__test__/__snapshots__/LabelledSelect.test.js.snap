// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelledSelect matches snapshot 1`] = `
<div>
  <label
    htmlFor="Species"
  >
    Species
  </label>
  <div
    className="css-1pcexqc-container"
    onKeyDown={[Function]}
  >
    <div
      className="css-12emboz"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="css-1hwfws3"
      >
        <div
          className="css-dvua67-singleValue"
        >
          Any
        </div>
        <input
          className="css-gj7qu5-dummyInput"
          disabled={false}
          id="Species-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          readOnly={true}
          tabIndex="0"
          value=""
        />
      </div>
      <div
        className="css-1wy0on6"
      >
        <span
          className="sc-bdVaJa jcfyVD"
          cx={[Function]}
        />
      </div>
    </div>
    <input
      name="Species"
      type="hidden"
      value=""
    />
  </div>
</div>
`;
